File: runme/mcgen/local/datan/solsrc/s1rn.c
#include <stdio.h>
#include <math.h>
#include "datsrc.h"
#include "grsrc.h"

extern void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                         double P_T *r, integer const P_T *n);

/* Table of constant values */

static integer c__5 = 5;
static double c_d1 = 1.;

main(void)
{
    /* Initialized data */
    static char tx[5] = "x    ", ty[5] = "N(x) ", capt[5] = "     ";
    /* Local variables */
    static integer i, n, iswit, nws, nx;
    static double a, delx, hist[50], r[1000], gamma, x0;
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S1RN demonstrates use of RNBW");
    printf("%s\n", "which is a subroutine generating random numbers");
    printf("%s\n\n", "following a Breit-Wigner distribution.");
/* let user choose */
    printf("%s\n", "Enter number N (<1001):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter mean value A:");
    printf("%s", "> ");
    scanf("%lf", &a);
    printf("%s\n", "Enter full width at half maximum GAMMA (>0.):");
    printf("%s", "> ");
    scanf("%lf", &gamma);
    printf("%s\n", "Do you want");
    printf("%s\n", "1 - numerical output");
    printf("%s\n", "2 - graphical output");
    printf("%s", "> ");
    scanf("%li", &iswit);
    rnbw(&a, &gamma, r, &n);
    if (iswit == 1) {
/* numerical output */
        for (i = 1; i <= n; ++i) {
            if (i % 5 == 1)
                printf(" ");
            printf("%12.6lf", r[i-1]);
            if (i % 5 == 0)
               printf("\n");
        }
        if (n % 5 != 0)
           printf("\n");
    } else if (iswit == 2) {
/* graphical output */
/* initialize histogram */
	x0 = a - gamma * 5.;
	delx = gamma * .2;
	nx = 50;
	smhsin_(hist, &x0, &delx, &nx);
/* fill histogram */
	i__1 = n;
	for (i = 1; i <= i__1; ++i) {
	    smhsfl_(hist, &x0, &delx, &nx, &r[i - 1], &c_d1);
	}
/* ask for number of workstation */
        grnbws_();
        printf("%s\n", "Please, enter number of workstation:");
        printf("%s", "> ");
        scanf("%li", &nws);
        (void)getchar();
/* display of histogram */
	smhsgr_(hist, &x0, &delx, &nx, tx, &c__5, ty, &c__5, capt, &c__5, &nws);
    }
    return 0;
} /* main */

/* ------------------------------------------------------------------ */
void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                  double P_T *r, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    static integer i;

/* a CALL of this subroutine generates in Array R a total of N */
/* random numbers following a Breit-Wigner distribution */
/* with mean A and full width at half maximum GAMMA */
    /* Parameter adjustments */
    --r;

    /* Function Body */
    rnecuy_(&r[1], n);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
	r[i] = *a + *gamma * .5 * tan((r[i] - .5) * 3.14159);
    }
    return;
} /* rnbw */

File: runme/mcgen/local/datan/solsrc/s1rn.f
      PROGRAM S1RN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION R(10000),HIST(50)
      CHARACTER*5 TX,TY,CAPT
      DATA TX/'x'/,TY/'N(x)'/,CAPT/' '/
C identify program to user
      WRITE(*,*)' Program S1RN demonstrates use of RNBW'
      WRITE(*,*)' which is a subroutine generating random numbers'
      WRITE(*,*)' following a Breit-Wigner distribution'
      WRITE(*,*)' '
C let user choose
      WRITE(*,*)' Enter number N (<10001)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter mean value A'
      WRITE(*,*)' >'
      READ(*,*)A
      WRITE(*,*)' Enter full width at half maximum GAMMA (>0.)'
      WRITE(*,*)' >'
      READ(*,*)GAMMA
      WRITE(*,*)' Do you want'
      WRITE(*,*)' 1 - numerical output'
      WRITE(*,*)' 2 - graphical output'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      CALL RNBW(A,GAMMA,R,N)
      IF(ISWIT.EQ.1) THEN
C numerical output
        WRITE(*,'(1X,5F12.6)')(R(I),I=1,N)
      ELSE IF(ISWIT.EQ.2) THEN
C graphical output
C initialize histogram
        X0=A-5.D0*GAMMA
        DELX=.2D0*GAMMA
        NX=50
        CALL SMHSIN(HIST,X0,DELX,NX)
C fill histogram
        DO 10 I=1,N
          CALL SMHSFL(HIST,X0,DELX,NX,R(I),1.D0)
10      CONTINUE
C ask for number of workstation
        CALL GRNBWS()
        WRITE(*,*)' Please, enter number of workstation:'
        WRITE(*,*)' >'
        READ(*,*)NWS
C display of histogram
        CALL SMHSGR(HIST,X0,DELX,NX,TX,5,TY,5,CAPT,5,NWS)
      END IF
      END
C ------------------------------------------------------------------
      SUBROUTINE RNBW(A,GAMMA,R,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(PI=3.14159D0,HALF=0.5D0)
C a CALL of this subroutine generates in Array R a total of N
C random numbers following a Breit-Wigner distribution
C with mean A and full width at half maximum GAMMA
      DIMENSION R(N)
      CALL RNECUY(R,N)
      DO 10 I=1,N
        R(I)=A+HALF*GAMMA*TAN(PI*(R(I)-HALF))
10    CONTINUE
      END
